packages/backend/src/inMemoryTableCalculations.ts

```diff
@@ -0,0 +1,127 @@
+import { DimensionType } from '@lightdash/common';
+import * as arrow from 'apache-arrow';
+import { Database } from 'duckdb-async';
+import Logger from './logging/logger';
+import { wrapOtelSpan } from './utils';
+
+type InMemoryJSONRow = Record<string, unknown>;
+type InMemoryJSONTableMap = Record<
+    string,
+    {
+        fields: Record<string, { type: DimensionType }>;
+        rows: InMemoryJSONRow[];
+    }
+>;
+
+export interface InMemoryDatabaseProvisioningOptions {
+    tables?: InMemoryJSONTableMap;
+}
+
+/**
+ * Converts DimensionType into an arrow data type. Arrow will gladly figure
+ * this out by itself, but it's prone to do it innacurately (e.g discarding
+ * time information on tiemestamps).
+ */
+function dimensionTypeToArrowDataType(
+    dimensionType: DimensionType,
+): arrow.DataType {
+    switch (dimensionType) {
+        case DimensionType.BOOLEAN:
+            return new arrow.Bool();
+        case DimensionType.DATE:
+            return new arrow.DateMillisecond();
+        case DimensionType.TIMESTAMP:
+            return new arrow.Timestamp(arrow.TimeUnit.MILLISECOND);
+        case DimensionType.NUMBER:
+            return new arrow.Float64();
+        case DimensionType.STRING:
+            return new arrow.Utf8();
+        default:
+            throw new Error(
+                `Unsupported dimension type to arrow datatype conversion: ${dimensionType}`,
+            );
+    }
+}
+
+/**
+ * Creates, and optionally provisions a new in-memory DuckDB instance.
+ *
+ * If table information is provided for provisioning, we use apache-arrow to load
+ * said table information into DuckDB, in the format { tableName: rows[] }
+ *
+ * We use a thin wrapper around DuckDB that exposes its methods as promises.
+ * See: https://www.npmjs.com/package/duckdb-async
+ */
+export const createDuckDbDatabase = async ({
+    tables,
+}: InMemoryDatabaseProvisioningOptions) => {
+    const db = await Database.create(':memory:');
+    await db.exec(`
+        INSTALL arrow;
+        LOAD arrow;
+
+        SET threads TO 1;
+        SET memory_limit = '512MB';
+        SET enable_external_access = false;
+    `);
+
+    if (tables) {
+        const loadedTableNames = await Promise.all(
+            Object.entries(tables).map(
+                async ([tableName, { rows, fields }]) => {
+                    /**
+                     * For each row, collect all values in all rows, and map the row type
+                     * into an arrow vector:
+                     */
+                    const columnVectors = Object.entries(fields).reduce(
```

Comment:
(optional) feels like this could be a hotspot since we're making copies with `vectorFromArray` and looping rows `N_field` number of times, could add more observability

Reply:
Good point, I'll open a follow-up PR adding a bit more measuring around this. In a future pass I still want to find ways to optimize the pipeline from warehouse->arrow->duckdb, but so far this was the best I could come up with.